tempesta_fw/sched/tfw_sched_http.c

```diff
@@ -85,46 +112,77 @@ MODULE_DESCRIPTION("Tempesta HTTP scheduler");
 MODULE_VERSION("0.3.1");
 MODULE_LICENSE("GPL");
 
-typedef struct {
-	TfwVhost *vhost;
-	TfwHttpMatchRule rule;
-} TfwSchedHttpRule;
+/* Active HTTP table. */
+static TfwHttpTable __rcu *tfw_table;
+/* Reconfig HTTP table. */
+static TfwHttpTable *tfw_table_reconfig;
+/* Entry for configuration particular HTTP chain of rules. */
+static TfwHttpChain *tfw_chain_entry;
+
+/*
+ * Scan all rules in linked chains of current active HTTP table. Main HTTP
+ * chain is processed primarily (must be first in the table list); if rule
+ * of some chain points to other chain - move to that chain and scan it.
+ */
+static TfwVhost *
+tfw_sched_http_table_scan(TfwMsg *msg, TfwHttpTable *table, bool *block)
+{
+	TfwHttpChain *chain;
+	TfwHttpMatchRule *rule;
+
+	chain = list_first_entry_or_null(&table->head, TfwHttpChain, list);
+	BUG_ON(!chain || chain->name);
+	while (chain) {
+		rule = tfw_http_match_req((TfwHttpReq *)msg, &chain->mark_list);
+		if (!rule)
+			rule = tfw_http_match_req((TfwHttpReq *)msg,
+						  &chain->match_list);
+		if (unlikely(!rule)) {
+			TFW_DBG("sched_http: No rule found in HTTP"
+				" chain '%s'\n", chain->name);
+			return NULL;
+		}
+		chain = (rule->act.type == TFW_HTTP_MATCH_ACT_CHAIN)
+		      ? rule->act.chain
+		      : NULL;
+	}
+
+	/* If rule points to virtual host, return the pointer. */
+	if (rule->act.type == TFW_HTTP_MATCH_ACT_VHOST)
+		return  rule->act.vhost;
 
-/* Active HTTP Scheduler rules. */
-static TfwHttpMatchList __rcu *tfw_rules;
-/* Reconfig HTTP Scheduler rules. */
-static TfwHttpMatchList *tfw_rules_reconfig;
+	/* If rule has 'block' action, request must be blocked. */
+	if (rule->act.type == TFW_HTTP_MATCH_ACT_BLOCK)
+		*block = true;
+
+	return NULL;
+}
 
 /*
  * Find vhost for an outgoing HTTP request.
  *
- * The search is based on contents of an HTTP request and match rules
- * that specify which Vhost the request should be forwarded to.
+ * The search is based on contents of an HTTP request and
+ * match http chain rules that specify which virtual host
+ * or other http chain the request should be forwarded to.
  */
 static TfwVhost *
-tfw_sched_http_sched_vhost(TfwMsg *msg)
+tfw_sched_http_sched_vhost(TfwMsg *msg, bool *block)
```

Comment:
For me no target vhost found means that request must be filtered. In current `http_chain` definition `block` is defined as default action. So rule `block` must be used if no target vhost found and variable `vhost` must be set to true. That means, that `block` variable duplicates return value. But the function works different: it doesn't set variable `block` to true, if no vhost found and no explicit `block` action encountered.

I see, why you've introduced `block` variable. Request without target vhost can be filtered in two ways: blocking of entire client connection or dropping a single request. I believe, HTTP scheduler is wrong place to make a that decision, it must be done in `http.c`. But let's look to the problem. Some clients may use proxies to reach TempestaFW, one client connection can multiplicate a lot of end clients. Some of them malicious, but most of them - not. We don't want to block good clients for others deeds, so dropping of single requests seems more reasonable for me. We can still block connections with lot of malicious requests using `http_resp_code_block` frang limit.

Reply:
It seems a request with not found vhost is dropped in `tfw_http_req_set_context()` with `cannot find vhost for request` and I think it has sense: a vhost isn't configured for the request - maybe it's just an error, not an attack. Anyway we do not pass the request (which also could be malicious) to a backend, so we're good.

Reply:
Done: connection dropped in case of `block` action; request dropped in case of `vhost` miss.